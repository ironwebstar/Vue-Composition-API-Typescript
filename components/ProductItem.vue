<template>
  <v-card class="mx-auto pa-5" max-width="344">
    <div class="wishlist--wrapper" cy-test="wishlist-action">
      <v-btn
        v-if="isInWishList"
        :key="wishList.length"
        rounded
        fab
        small
        color="white"
        @click="removeWishList(product)"
      >
        <v-icon color="red">mdi-star</v-icon>
      </v-btn>
      <v-btn
        v-else
        :key="wishList.length"
        fab
        small
        rounded
        @click="addWishList(product)"
      >
        <v-icon color="grey lighten-2">mdi-star</v-icon>
      </v-btn>
    </div>

    <v-img :src="product.cover_image_url" height="200px" />

    <v-card-title class="block-ellipsis" cy-test="product-title">
      {{ product.title }}
    </v-card-title>

    <v-card-text>
      <v-row align="center" class="mx-0">
        <v-rating
          :value="product.reviews_avg"
          color="amber"
          dense
          half-increments
          readonly
          size="14"
          class="white"
        ></v-rating>

        <div class="grey--text ml-4">
          {{ `${product.reviews_avg} (${product.reviews_number})` }}
        </div>
      </v-row>
      <v-row
        align="center"
        class="text-body-1 font-weight-bold mx-0 pt-3 text-center"
      >
        <div
          v-if="product.discount !== 0"
          class="text-decoration-line-through pr-3"
        >
          {{ product.retail_price.formatted_value }}
        </div>
        <div class="red--text">{{ product.retail_price.formatted_value }}</div>
      </v-row>
    </v-card-text>

    <v-card-subtitle cy-test="product-description" class="block-ellipsis">
      {{ product.description }}
    </v-card-subtitle>

    <v-card-actions :key="`${cart.length}`" cy-test="cart-action">
      <v-btn
        v-if="isInCart"
        text
        outlined
        block
        rounded
        color="black"
        class="text-uppercase grey lighten-2 remove-cart__btn"
        @click="removeCart(product)"
      >
        In cart
      </v-btn>
      <v-hover v-else v-slot="{ hover }">
        <v-btn
          text
          block
          rounded
          outlined
          class="text-uppercase add-cart__btn"
          :color="hover ? 'white' : 'black'"
          :class="hover ? 'black' : ''"
          @click="addCart(product)"
        >
          Add to cart
        </v-btn>
      </v-hover>
    </v-card-actions>
  </v-card>
</template>

<script lang="ts">
import { defineComponent, computed, ComputedRef } from '@nuxtjs/composition-api'
import { ProductType } from '~/interfaces'
import productCompose from '~/composables/product'

export default defineComponent({
  name: 'ProductItem',

  props: {
    product: {
      type: Object as () => ProductType,
      required: true,
    },
  },

  setup(props) {
    const {
      cart,
      wishList,
      addCart,
      removeCart,
      addWishList,
      removeWishList,
    } = productCompose()

    const isInCart: ComputedRef<Boolean> = computed((): boolean => {
      return cart.value.some((e: ProductType) => e.uuid === props.product.uuid)
    })

    const isInWishList: ComputedRef<Boolean> = computed((): boolean => {
      return wishList.value.some(
        (e: ProductType) => e.uuid === props.product.uuid
      )
    })

    return {
      // computed
      cart,
      wishList,
      // methods
      addCart,
      removeCart,
      addWishList,
      removeWishList,
      isInCart,
      isInWishList,
    }
  },
})
</script>

<style lang="scss" scoped>
.wishlist--wrapper {
  position: absolute;
  z-index: 1;
  right: 5px;
  top: 5px;
}
</style>
