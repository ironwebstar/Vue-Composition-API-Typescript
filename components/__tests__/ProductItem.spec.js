import Vuex from 'vuex'
import Vuetify from 'vuetify'
import { mount, createLocalVue } from '@vue/test-utils'

import ProductItem from '@/components/ProductItem.vue'
import { productValue } from '~/test/unit/mock-input'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('ProductItem', () => {
  let actions
  let store
  let mutations
  let getters
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
    actions = {
      getProducts: jest.fn(),
    }
    mutations = {
      ADD_CART: jest.fn(),
      REMOVE_CART: jest.fn(),
      ADD_WISHLIST: jest.fn(),
      REMOVE_WISHLIST: jest.fn(),
      SET_PRODUCTS: jest.fn(),
    }
    getters = {
      products: () => [productValue],
      cart: () => [],
      wishList: () => [],
    }

    store = new Vuex.Store({
      actions,
      mutations,
      getters,
    })
  })

  it('should be a Vue instance', () => {
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })
    expect(wrapper.vm).toBeTruthy()
  })

  it('should be render title, description name', () => {
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })
    const title = wrapper.find('.v-card__title')
    expect(title.text()).toBe('Title')
    const description = wrapper.find('.v-card__subtitle')
    expect(description.text()).toBe('Description')
  })

  it('should call addCart method when click button', () => {
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })

    wrapper.vm.addCart = jest.fn()
    const addToCartBtn = wrapper.find('.v-card__actions .add-cart__btn')
    expect(addToCartBtn.text().toUpperCase()).toBe('ADD TO CART')

    addToCartBtn.trigger('click')
    expect(wrapper.vm.addCart).toHaveBeenCalledTimes(1)
  })

  it('should call removeCart method when click button', async () => {
    getters = {
      products: () => [productValue],
      cart: () => [],
      wishList: () => [],
    }

    store = new Vuex.Store({
      actions,
      mutations,
      getters,
    })
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })
    const removeCartBtn = wrapper.find('.v-card__actions button')

    await removeCartBtn.trigger('click')
    expect(wrapper.find('.v-card__actions').text().toUpperCase()).toBe(
      'IN CART'
    )
  })

  it('should call addWishList method when click button', () => {
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })

    wrapper.vm.addWishList = jest.fn()
    const addWishListBtn = wrapper.find('.wishlist--wrapper button')

    addWishListBtn.trigger('click')
    expect(wrapper.vm.addWishList).toHaveBeenCalledTimes(1)
  })

  it('should call removeWishList method when click button', async () => {
    getters = {
      products: () => [productValue],
      cart: () => [],
      wishList: () => [productValue],
    }

    store = new Vuex.Store({
      actions,
      mutations,
      getters,
    })
    const wrapper = mount(ProductItem, {
      localVue,
      vuetify,
      store,
      propsData: {
        product: productValue,
      },
    })
    const removeWishListBtn = wrapper.find('.wishlist--wrapper button')

    await removeWishListBtn.trigger('click')
    expect(wrapper.find('i').classes()).toContain('red--text')
  })
})
