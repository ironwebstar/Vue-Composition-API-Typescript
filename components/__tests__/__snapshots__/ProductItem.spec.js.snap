// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem should be a Vue instance 1`] = `
"<div class=\\"mx-auto pa-5 v-card v-sheet theme--light\\" style=\\"max-width: 344px;\\">
  <div class=\\"wishlist--wrapper\\"><button type=\\"button\\" class=\\"v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round v-btn--rounded theme--light v-size--small white\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-star theme--light red--text\\"></i></span></button></div>
  <div class=\\"v-image v-responsive theme--light\\" style=\\"height: 200px;\\">
    <transition-stub name=\\"fade-transition\\" mode=\\"in-out\\">
      <div class=\\"v-image__image v-image__image--preload v-image__image--cover\\" style=\\"background-position: center center;\\"></div>
    </transition-stub>
    <div class=\\"v-responsive__content\\"></div>
  </div>
  <div class=\\"v-card__title\\">Title</div>
  <div class=\\"v-card__text\\">
    <div class=\\"row mx-0 align-center\\">
      <div class=\\"v-rating v-rating--readonly v-rating--dense\\"><button type=\\"button\\" tabindex=\\"-1\\" aria-label=\\"Rating 1 of 5\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star theme--light amber--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" aria-label=\\"Rating 2 of 5\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star theme--light amber--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" aria-label=\\"Rating 3 of 5\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star theme--light amber--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" aria-label=\\"Rating 4 of 5\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star theme--light amber--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" aria-label=\\"Rating 5 of 5\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-half-full theme--light amber--text\\" style=\\"font-size: 14px;\\"></button></div>
      <div class=\\"grey--text ml-4\\">
        4.6 (17)
      </div>
    </div>
  </div>
  <div class=\\"v-card__subtitle\\">Description</div>
  <div class=\\"v-card__actions\\"><button type=\\"button\\" class=\\"text-uppercase grey lighten-2 v-btn v-btn--block v-btn--outlined v-btn--rounded v-btn--text theme--light v-size--default black--text\\"><span class=\\"v-btn__content\\">
      In cart
    </span></button></div>
</div>"
`;
